usecase/final_characters_usecase.go

```diff
@@ -0,0 +1,87 @@
+package usecase
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/SergioSanchezWizeline/academy-go-q12021/domain/model"
+	"github.com/SergioSanchezWizeline/academy-go-q12021/interfaces"
+)
+
+type finalCharacters struct {
+	Repository interfaces.CharacterRepository
+}
+
+type FinalCharactersUserCase interface {
+	Execute(opType string, items int, itemsPerWorker int) ([]*model.Character, error)
+}
+
+func NewFinalCharactersUseCase(repository interfaces.CharacterRepository) FinalCharactersUserCase {
+	return &finalCharacters{repository}
+}
+
+func (usecase *finalCharacters) Execute(opType string, items int, itemsPerWorker int) ([]*model.Character, error) {
+
+	if opType != "odd" && opType != "even" {
+		return nil, fmt.Errorf("invalid parameter value for 'type': %v valid values are 'even' and 'odd'", opType)
+	}
+
+	if items == 0 || itemsPerWorker == 0 {
+		return []*model.Character{}, nil
+	}
+
+	all, err := usecase.Repository.All()
+	if err != nil {
+		return nil, err
+	}
+
+	poolSize := items/itemsPerWorker + 1
+	jobs := make(chan *model.Character, poolSize)
+	result := make(chan *model.Character, items)
+	safeCounter := struct {
+		mu    sync.Mutex
+		count int
+	}{count: poolSize}
```

Comment:
An alternative for this solution would be to replace this `safeCounter` implementation by using an atomic operation whenever you need to update the counter.

Actually, it'd be interesting to see which approach would be more performant. Maybe it's a good excuse to practice [benchmarks](https://golang.org/pkg/testing/#hdr-Benchmarks). 😛

Reply:
Done